File: app/api/process-document/route.ts
import { NextRequest, NextResponse } from 'next/server';
import pdfParse from 'pdf-parse';
import Tesseract from 'tesseract.js';
import fetch from 'node-fetch';
import * as cheerio from 'cheerio';

export const runtime = 'nodejs';

async function extractTextFromPDF(buffer: Buffer): Promise<string> {
  const data = await pdfParse(buffer);
  return data.text;
}

async function extractTextFromImage(buffer: Buffer): Promise<string> {
  const { data: { text } } = await Tesseract.recognize(buffer, 'tur');
  return text;
}

async function extractTextFromURL(url: string): Promise<string> {
  try {
    const res = await fetch(url);
    const html = await res.text();
    const $ = cheerio.load(html);
    const text = $('p,li').map((_, el) => $(el).text()).get().join('\n');
    return text;
  } catch (e) {
    return `[URL'den metin çekilemedi: ${e}]`;
  }
}

function detectFileType(buffer: Buffer): 'pdf' | 'image' | 'unknown' {
  if (buffer.slice(0, 4).toString() === '%PDF') return 'pdf';
  if (buffer.slice(0, 8).toString('hex') === '89504e470d0a1a0a') return 'image'; // PNG
  if (buffer.slice(0, 2).toString('hex') === 'ffd8') return 'image'; // JPG
  return 'unknown';
}

function getPrompt(processType: string, text: string): string {
  if (processType === 'contract-analyzer')
    return `Aşağıdaki sözleşmenin en önemli maddelerini ve risklerini madde madde özetle:\n\n${text}`;
  if (processType === 'meeting-summarizer')
    return `Aşağıdaki toplantı notlarını kısa ve anlaşılır şekilde özetle:\n\n${text}`;
  if (processType === 'document-summarizer')
    return `Aşağıdaki metni (kitap, makale veya belge) kısa ve öz şekilde özetle:\n\n${text}`;
  if (processType === 'setup-guide-extractor')
    return `Aşağıdaki metinden kullanım veya kurulum adımlarını madde madde çıkar:\n\n${text}`;
  return `Aşağıdaki metni özetle:\n\n${text}`;
}

async function callGeminiLLM(prompt: string): Promise<string> {
  const apiKey = process.env.GOOGLE_API_KEY;
  if (!apiKey) return '[GOOGLE_API_KEY ortam değişkeni tanımlı değil!]';
  const res = await fetch('https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent?key=' + apiKey, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ contents: [{ parts: [{ text: prompt }] }] })
  });
  const data = await res.json();
  return data?.candidates?.[0]?.content?.parts?.[0]?.text || '[LLM cevabı alınamadı]';
}

export async function POST(req: NextRequest) {
  try {
    const data = await req.json();
    const { file_content_base64, url, text_content, process_type } = data;
    let extracted_text = '';
    if (file_content_base64) {
      const buffer = Buffer.from(file_content_base64, 'base64');
      const fileType = detectFileType(buffer);
      if (fileType === 'pdf') {
        extracted_text = await extractTextFromPDF(buffer);
      } else if (fileType === 'image') {
        extracted_text = await extractTextFromImage(buffer);
      } else {
        extracted_text = '[Desteklenmeyen dosya türü veya metin çıkarılamadı.]';
      }
    } else if (url) {
      extracted_text = await extractTextFromURL(url);
    } else if (text_content) {
      extracted_text = text_content;
    } else {
      return NextResponse.json({ error: 'Geçerli bir giriş bulunamadı.' }, { status: 400 });
    }
    const prompt = getPrompt(process_type, extracted_text);
    const llm_result = await callGeminiLLM(prompt);
    return NextResponse.json({ extracted_text, llm_result });
  } catch (e: any) {
    return NextResponse.json({ error: e.message }, { status: 500 });
  }
} 

File: app/api/process-document/route.py
import base64
import os
import tempfile
from fastapi import Request
from fastapi.responses import JSONResponse
from vercel_python import VercelRequest, VercelResponse, VercelFastAPI

# PDF için
import fitz  # PyMuPDF
# Resim için
from PIL import Image
import pytesseract
# Web sayfası için
import requests
from bs4 import BeautifulSoup

# Google Gemini
import google.generativeai as genai

app = VercelFastAPI()

def extract_text_from_pdf(file_bytes):
    with tempfile.NamedTemporaryFile(delete=False, suffix=".pdf") as tmp_file:
        tmp_file.write(file_bytes)
        tmp_file.flush()
        doc = fitz.open(tmp_file.name)
        text = "\n".join([page.get_text() for page in doc])
        doc.close()
    os.unlink(tmp_file.name)
    return text

def extract_text_from_image(file_bytes):
    with tempfile.NamedTemporaryFile(delete=False, suffix=".png") as tmp_file:
        tmp_file.write(file_bytes)
        tmp_file.flush()
        image = Image.open(tmp_file.name)
        text = pytesseract.image_to_string(image, lang="tur")
    os.unlink(tmp_file.name)
    return text

def extract_text_from_url(url):
    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        soup = BeautifulSoup(response.text, "html.parser")
        paragraphs = soup.find_all(['p', 'li'])
        text = "\n".join([p.get_text() for p in paragraphs])
        return text
    except Exception as e:
        return f"[URL'den metin çekilemedi: {str(e)}]"

def detect_file_type(file_bytes):
    if file_bytes[:4] == b"%PDF":
        return "pdf"
    if file_bytes[:8] == b"\x89PNG\r\n\x1a\n":
        return "image"
    if file_bytes[:2] == b"\xff\xd8":
        return "image"
    return "unknown"

def get_prompt(process_type, text):
    if process_type == "contract-analyzer":
        return f"Aşağıdaki sözleşmenin en önemli maddelerini ve risklerini madde madde özetle:\n\n{text}"
    elif process_type == "meeting-summarizer":
        return f"Aşağıdaki toplantı notlarını kısa ve anlaşılır şekilde özetle:\n\n{text}"
    elif process_type == "document-summarizer":
        return f"Aşağıdaki metni (kitap, makale veya belge) kısa ve öz şekilde özetle:\n\n{text}"
    elif process_type == "setup-guide-extractor":
        return f"Aşağıdaki metinden kullanım veya kurulum adımlarını madde madde çıkar:\n\n{text}"
    else:
        return f"Aşağıdaki metni özetle:\n\n{text}"

def call_gemini_llm(prompt):
    api_key = os.environ.get("GOOGLE_API_KEY")
    if not api_key:
        return "[GOOGLE_API_KEY ortam değişkeni tanımlı değil!]"
    genai.configure(api_key=api_key)
    model = genai.GenerativeModel("gemini-pro")
    response = model.generate_content(prompt)
    return response.text if hasattr(response, 'text') else str(response)

@app.post("")
async def process_document(request: Request):
    try:
        data = await request.json()
        file_content_base64 = data.get("file_content_base64")
        url = data.get("url")
        text_content = data.get("text_content")
        process_type = data.get("process_type")

        extracted_text = None
        file_type = None

        if file_content_base64:
            file_bytes = base64.b64decode(file_content_base64)
            file_type = detect_file_type(file_bytes)
            if file_type == "pdf":
                extracted_text = extract_text_from_pdf(file_bytes)
            elif file_type == "image":
                extracted_text = extract_text_from_image(file_bytes)
            else:
                extracted_text = "[Desteklenmeyen dosya türü veya metin çıkarılamadı.]"
        elif url:
            extracted_text = extract_text_from_url(url)
        elif text_content:
            extracted_text = text_content
        else:
            return JSONResponse({"error": "Geçerli bir giriş bulunamadı."}, status_code=400)

        prompt = get_prompt(process_type, extracted_text)
        llm_result = call_gemini_llm(prompt)

        result = {
            "extracted_text": extracted_text,
            "llm_result": llm_result
        }
        return JSONResponse(result)
    except Exception as e:
        return JSONResponse({"error": str(e)}, status_code=500) 